core/src/main/java/io/grpc/ForwardingChannelBuilder.java

```diff
@@ -139,6 +139,24 @@ public T maxInboundMessageSize(int max) {
     return thisT();
   }
 
+  @Override
```

Comment:
An API Alternative: this might be cleaner as some sort of "connectionConfig"  where a user can set each of the three options in a single object.   There are a couple connection specific options that a user might want to set, and it would be nice to not have a method for each one.  Max concurrent streams would be one option.  Another one would be to eagerly create connections, something that the channel state API handles now.

Reply:
That's a substantially different API, and I don't think I'd feel comfortable with it. It would not be extensible on a per-transport basis and we'd be limited to ignoring unknown configuration. I guess it could be a sub-builder for per-transport configuration, but I'm not sure what the added value is. Also, for the user, the distinction between "channel" and "connection" is nebulous.

Max concurrent streams is only on server-side, although that doesn't mean we couldn't mirror the API here. The channel state API was designed to be the place that would trigger eager connection creation, so I'm not excited to adding something equivalent here.

Reply:
I know its a different API, though it doesn't seem inherently worse.  The configuration object could be extensible, with each transport providing a child class.  This is the netty pattern, and it works pretty well.

The API in this PR is adding three additional methods that most users will not need to set.  I think we should be careful to keep the MCB API clean and trim.

Reply:
> This is the netty pattern, and it works pretty well.

Netty's still a bit different than that because bootstrap has all concrete classes. You have the option to pass an Option to it and you can also change the channel-specific Config after Channel creation.

Going with this approach would have left ManagedChannelBuilder concrete with nobody extending it. While I'm not arguing whether that may have been better, it seems a stark departure to what we have now and unnecessary given we are already extending ManagedChannelBuilder. If you're proposing we get rid of extending ManagedChannelBuilder, I guess that may be possible, but it'd be quite a redesign.

Reply:
The reason I am thinking about alternatives is because the API you are adding here is not experimental (and not sure it can be).   I'd be less concerned if this API change wasn't permanent, but since it is I think we need to evaluate alternatives before proceeding.